modules/patching/diff.go

```diff
@@ -0,0 +1,275 @@
+}
+
+// ConvertToCRLF converts this diff from using LF to CRLF line separators given the base text to patch.
+// TODO(wongb): Review this for performance
+func (diff *Diff) ConvertToCRLF(base string) *Diff {
+	newStartIndex := diff.StartIndex
+	newChanges := strings.Replace(diff.Changes, "\n", "\r\n", -1)
+
+	for i := 0; i < newStartIndex && i < len(base)-1; i++ {
+		if base[i] == '\r' && base[i+1] == '\n' {
+			newStartIndex++
+		}
+	}
+
+	return NewDiff(diff.Insertion, newStartIndex, newChanges)
+}
+
+// ConvertToLF converts this diff from using CRLF to LF line separators given the base (CRLF) text that it was generated from.
+// TODO(wongb): Review this for performance
+func (diff *Diff) ConvertToLF(base string) *Diff {
+	newStartIndex := diff.StartIndex
+	newChanges := strings.Replace(diff.Changes, "\r\n", "\n", -1)
+
+	for i := 0; i < diff.StartIndex-1 && i < len(base)-1; i++ {
+		if base[i] == '\r' && base[i+1] == '\n' {
+			newStartIndex--
+		}
+	}
+
+	return NewDiff(diff.Insertion, newStartIndex, newChanges)
+}
+
+// Undo reverses this diff, producing a diff to undo the changes done by applying the diff.
+func (diff *Diff) Undo() *Diff {
+	return NewDiff(!diff.Insertion, diff.StartIndex, diff.Changes)
+}
+
+// OffsetDiff shifts the start index of this diff by the provided offset
+func (diff *Diff) OffsetDiff(offset int) *Diff {
+	return NewDiff(diff.Insertion, diff.StartIndex+offset, diff.Changes)
+}
+
+func (diff *Diff) subChanges(start, end int) *Diff {
+	return NewDiff(diff.Insertion, diff.StartIndex, diff.Changes[start:end])
+}
+
+func (diff *Diff) subChangesStartingFrom(start int) *Diff {
+	return NewDiff(diff.Insertion, diff.StartIndex, diff.Changes[start:])
+}
+
+func (diff *Diff) subChangesEndingAt(end int) *Diff {
+	return NewDiff(diff.Insertion, diff.StartIndex, diff.Changes[:end])
+}
+
+func (diff *Diff) transform(others []*Diff) []*Diff {
+	intermediateDiffs := []*Diff{}
+	intermediateDiffs = append(intermediateDiffs, diff)
+
+	for _, other := range others {
+		newIntermediateDiffs := []*Diff{}
+		for _, current := range intermediateDiffs {
+			switch {
+			// CASE 1: IndexA < IndexB
+			case other.StartIndex < current.StartIndex:
+				switch {
+				// CASE 1a, 1b: Ins - Ins, Ins - Rmv
+				case other.Insertion && current.Insertion, other.Insertion && !current.Insertion:
+					newIntermediateDiffs = doTransform(transformType2, newIntermediateDiffs, current, other)
+				// CASE 1c: Rmv - Ins
+				case !other.Insertion && current.Insertion:
+					newIntermediateDiffs = doTransform(transformType3, newIntermediateDiffs, current, other)
+				// CASE 1d: Rmv - Rmv
+				case !other.Insertion && !current.Insertion:
+					newIntermediateDiffs = doTransform(transformType4, newIntermediateDiffs, current, other)
+				// Fail; should never have gotten to here.
+				default:
+					panic(fmt.Sprintf("Got to invalid state 1e while transforming [%s] on predessor [%+v], from list [%+v]", current.String(), other, others))
```

Comment:
Same comment here about parsing arguments into errors, though this is slightly different. I think it makes sense to have the state name in the error, but the patch info isn't exactly necessary to throw.
Also, is it right to be panicking here? do we really want an error applying a patch to potentially take down the server if it's not caught? Or rather, if we are going to catch it, why not just have it return an error?